File: Chapter01/python/packt1/activity/zipf_validation.py
from pyspark.rdd import RDD
from pyspark.sql import SparkSession
import re
import os
from operator import add

from Chapter01.python.packt1.helper_python import create_session, novella_location

if __name__ == "__main__":
    session: SparkSession = create_session(2, "Zipf validation")
    lines: RDD = session.sparkContext.textFile(novella_location)

    counts_per_token = lines \
        .flatMap(lambda line: line.lower().split()) \
        .map(lambda word: (re.sub(r"(^[^a-z0-9]+|[^a-z0-9]+$)", '', word), 1)) \
        .reduceByKey(add)

    sorted_counts = counts_per_token.map(lambda tuple: (tuple[1], tuple[0])) \
        .sortByKey(ascending=False)

    sorted_counts.saveAsTextFile(os.path.join('.', 'sortedcounts'))

File: Chapter01/scala/packt1/activity/ZipfValidation.scala
package packt1.solutions

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import packt1.HelperScala.{createSession, separator, novellaLocation}

object ZipfValidation {

  def main(args: Array[String]): Unit = {

    val session: SparkSession = createSession(2, "Zipf Validation")
    val lines: RDD[String] = session.sparkContext.textFile(novellaLocation)

    val countsPerToken: RDD[(String, Int)] = lines
      .flatMap(_.toLowerCase.split("\\s+")) // split on whitespace
      .map(word => (word.replaceAll("(^[^a-z0-9]+|[^a-z0-9]+$)", ""), 1)) // removing punctuation
      .reduceByKey((count1, count2) => count1 + count2, 2)

    val sortedCountsToken = countsPerToken.map(_.swap) // swap is equivalent to `pair =>(pair._2, pair._1)`
      .sortByKey(ascending=false)

    sortedCountsToken.saveAsTextFile(s"Chapter01${separator}sortedcounts")
  }

}
